p2p/security/noise/handshake.go

```diff
@@ -102,12 +97,22 @@ func (s *secureSession) runHandshake(ctx context.Context) (err error) {
 		if err != nil {
 			return fmt.Errorf("error reading handshake message: %w", err)
 		}
-		if err := s.handleRemoteHandshakePayload(plaintext, hs.PeerStatic()); err != nil {
+		rcvdEd, err := s.handleRemoteHandshakePayload(plaintext, hs.PeerStatic())
+		if err != nil {
 			return err
 		}
+		if s.earlyDataHandler != nil {
+			if err := s.earlyDataHandler.Received(ctx, s.insecureConn, rcvdEd); err != nil {
```

Comment:
The early data handler may need to know the handshake state to drive its internal state machine, for example if this is the server or client side, at which stage is the early data received. Do you think we should pass the handshake state info to the earlyDataHandler too?  or maybe that is available in other ways already and I missed it?

Reply:
For both client and server, the data is sent with their respective first flights.

I think the nicest way to resolve this is to pass separate client and server `EarlyDataHandlers` to the `EarlyData` option. I applied that change in 0b3683270e9380124b675625cf77f0113f7905e4. What do you think?

Reply:
That looks great, thanks for the changes which make the Noise based muxer selection and other early data utilization smoother.